### PR TITLE
Revise flow for adding/removing program admins

### LINKED ISSUE #4876: Manage Admins UI doesn't make it clear what info is needed
**Is your feature request related to a problem? Please describe.**
The "Manage Admins" UI doesn't make it clear that the text field is for email addresses. We should improve this UX:

Screenshot 2023-05-16 at 9 27 04 PM

**Describe the solution you'd like**

**Additional context**

**Done when**
What conditions need to be met before this issue can be closed?

### PR DESCRIPTION
**Pull request by yotommy**
### Description

Revise flow for adding/removing program admins via email address.

## Release notes

Revise flow for adding/removing program admins via email address. The new flow adds or deletes a single admin per action and uses standard UI elements.

#### General

Read the full guidelines for PRs [here](https://docs.civiform.us/contributor-guide/developer-guide/technical-contribution-guide#guidelines-for-pull-requests)

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release | database >
- [x] Created unit and/or browser tests which fail without the change (if possible)
- [x] Performed manual testing (Chrome and Firefox if it includes front-end changes)

### Instructions for manual testing

- Navigate to Program Dashboard
- Under three-dot menu, choose "Manage Program Admins"
- Add and delete admins. Accounts must exist and the accounts cannot be CiviForm Admins
- Click "Back" to return to Program Dashboad
- Under three-dot menu, choose "Manage Program Admins" to ensure the list has been properly persisted.

### Issue(s) this completes

Fixes #4876

image

### CODE REVIEW

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 1]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
     }
   }
 
+  /** Deletes an existing admin email. */
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:88]: Nit: can you rephrase this to clarify that it removes a user as a program admin for the specified program? Current phrasing could be read as deleting an email message or resetting an account email, for example.

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 2]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
...
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result delete(Http.Request request, long programId, String adminEmail) {
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:90]: There should be new routes for these new endpoints

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 3]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
...
+    try {
+      roleService.removeProgramAdmins(programId, ImmutableSet.of(adminEmail));
+      return redirect(routes.AdminProgramController.index());
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:93]: This will redirect to the program list. Should it redirect back to the program admin management page instead?

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 4]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
...
+
+  /** Adds a new admin email. */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result add(Http.Request request, long programId, String adminEmail) {
+    try {
+      Optional<CiviFormError> maybeError =
+          roleService.makeProgramAdmins(programId, ImmutableSet.of(adminEmail));
+
+      if (maybeError.isEmpty()) {
+        return redirect(routes.AdminProgramController.index());
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:107]: Same comment here re: redirect location. You can call loadProgram(request, programId, Optional.empty()) instead.

File: server/test/controllers/admin/ProgramAdminManagementControllerTest.java [hunk 5]
```diff
@@ -163,4 +163,65 @@ public void update_programNotFound_returnsNotFound() {
 
     assertThat(result.status()).isEqualTo(NOT_FOUND);
   }
+
+  @Test
+  public void add_succeeds() {
```
Comment (reemaka) [inline server/test/controllers/admin/ProgramAdminManagementControllerTest.java:168]: Can you test the error case?

Comment (yotommy) [reply server/test/controllers/admin/ProgramAdminManagementControllerTest.java:168]: I'm not sure how to test this. I tried writing one for the case where `makeProgramAdmins()` returns `Optional.empty`, but this ends up just testing that the return code is `OK` instead of `SEE_OTHER`. What do you think?

Comment (yotommy) [reply server/test/controllers/admin/ProgramAdminManagementControllerTest.java:168]: I looked through our tests to see if there is a way of asserting something about the `Toast` but did not find anything. I certainly may have missed something, though.

File: server/conf/routes [hunk 6]
```diff
@@ -29,8 +29,10 @@ POST    /admin/programs/:programId                              controllers.admi
 POST    /admin/programs/:programId/setEligibilityIsGating       controllers.admin.AdminProgramController.setEligibilityIsGating(request: Request, programId: Long)
 
 # Routes for a CiviForm admin to manage program admins for a given program.
-GET     /admin/programs/:programId/admins/edit  controllers.admin.ProgramAdminManagementController.edit(request: Request, programId: Long)
-POST    /admin/programs/:programId/admins       controllers.admin.ProgramAdminManagementController.update(request: Request, programId: Long)
+GET     /admin/programs/:programId/admins/edit    controllers.admin.ProgramAdminManagementController.edit(request: Request, programId: Long)
+POST    /admin/programs/:programId/admins/delete  controllers.admin.ProgramAdminManagementController.delete(request: Request, programId: Long, adminEmail: String)
```
Comment (reemaka) [inline server/conf/routes:33]: adminEmail needs to be in the route, i.e.: /admin/programs/:programId/admins/:adminEmail/delete

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 7]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
     }
   }
 
+  /** Removes `adminEmail` as a program admin for the program identified by `programId`. */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result delete(Http.Request request, long programId, String adminEmail) {
+    try {
+      roleService.removeProgramAdmins(programId, ImmutableSet.of(adminEmail));
+      return redirect(routes.AdminProgramController.edit(programId));
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:93]: This will redirect to the program edit view. I think you want:
```
return ok(routes.ProgramAdminManagementController.edit(programId);
```

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 8]
```diff
@@ -85,6 +85,36 @@ public Result update(Http.Request request, long programId) {
...
+
+  /** Adds a new admin email. */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result add(Http.Request request, long programId, String adminEmail) {
+    try {
+      Optional<CiviFormError> maybeError =
+          roleService.makeProgramAdmins(programId, ImmutableSet.of(adminEmail));
+
+      if (maybeError.isEmpty()) {
+        return redirect(routes.AdminProgramController.edit(programId));
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:107]: Same here

File: server/test/controllers/admin/ProgramAdminManagementControllerTest.java (hunk 5 above)
Comment (reemaka) [reply server/test/controllers/admin/ProgramAdminManagementControllerTest.java:168]: I think you can look at the returned html in the response, but maybe not necessary. Fine to leave it as is in that case, thanks for checking!

File: server/app/controllers/admin/ProgramAdminManagementController.java (hunk 7 above)
Comment (yotommy) [reply server/app/controllers/admin/ProgramAdminManagementController.java:93]: As discussed via Slack, redirecting to the new (correct) view that you identified.

File: server/app/controllers/admin/ProgramAdminManagementController.java (hunk 8 above)
Comment (yotommy) [reply server/app/controllers/admin/ProgramAdminManagementController.java:107]: As above.

File: server/conf/routes (hunk 6 above)
Comment (yotommy) [reply server/conf/routes:33]: Done.

File: browser-test/image_snapshots/admin_program_admins_test/add-program-admin-error.png
Comment (reemaka) [inline browser-test/image_snapshots/admin_program_admins_test/add-program-admin-error.png:1]: Could we add a screenshot of when there are existing admins?

File: server/app/controllers/admin/ProgramAdminManagementController.java [hunk 9]
```diff
@@ -52,34 +52,39 @@ public Result edit(Http.Request request, long programId) {
...
     }
-    ManageProgramAdminsForm manageAdminForm = form.bindFromRequest(request).get();
+  }
 
+  /** Adds a new admin email. */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result add(Http.Request request, long programId) {
+    Form<ManageProgramAdminsForm> form =
+        formFactory.form(ManageProgramAdminsForm.class).bindFromRequest(request);
+    if (Strings.isNullOrEmpty(form.get().getAdminEmail())) {
```
Comment (reemaka) [inline server/app/controllers/admin/ProgramAdminManagementController.java:73]: I think we should trim the email in case there's whitespace.

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 10]
```diff
@@ -64,72 +71,105 @@ public Content render(
             .setTitle(fullTitle)
             .addMainContent(
                 renderHeader(fullTitle),
-                adminEmailTemplate(),
-                renderAdminForm(request, program.id(), existingAdminEmails));
+                renderHeader("Add New Admin"),
+                renderAddNewButton(request, program.id()),
+                renderHeader("Existing Admins"),
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:76]: renderHeader adds an h1 tag and there should only be one h1 tag per page. Instead, you probably want to use h2() for "Add New Admin" and "Existing Admins"

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 11]
```diff
@@ -64,72 +71,105 @@ public Content render(
...
 
     return layout.renderCentered(htmlBundle);
   }
 
-  /**
-   * Render a form with inputs for program admin emails. If program admins exist, the input fields
-   * will be pre-populated with their email addresses.
-   */
-  private FormTag renderAdminForm(
+  private DivTag renderAddNewButton(Http.Request request, long programId) {
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:85]: Maybe renderAddNewAdminForm rather than Button since it returns the whole form and not just the button?

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 12]
```diff
@@ -64,72 +71,105 @@ public Content render(
...
+  }
+
+  private DivTag renderExistingAdmins(
       Http.Request request, long programId, ImmutableList<String> existingAdminEmails) {
-    DivTag emailFields =
-        div()
-            .withId(EMAIL_CONTAINER_DIV_ID)
-            .withClasses("ml-4")
-            .with(each(existingAdminEmails, email -> adminEmailInput(Optional.of(email))))
+    return div(
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:108]: optional: I think you can do `div().with(table() ....);` instead of passing the table to the div and that might be more readable.

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 13]
```diff
@@ -64,72 +71,105 @@ public Content render(
...
+            .with(
+                th("Actions")
+                    .withClasses(BaseStyles.TABLE_CELL_STYLES, "text-right", "pr-8", "w-1/3")));
   }
 
-  private DivTag adminEmailInput(Optional<String> existing) {
-    // When there are existing admins, the only option is to remove that admin. The field is
-    // disabled, so that no changes except removal can be made. The form does not submit disabled
-    // fields, so these existing admins will not be removed unless the remove button is clicked,
-    // which sets disabled to false (see TypeScript file).
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:104]: Can the typescript code be removed or is it still being used?

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 14]
```diff
@@ -64,72 +71,105 @@ public Content render(
...
-    String inputFieldName = existing.isPresent() ? REMOVE_EMAIL_FIELD_NAME : ADD_EMAIL_FIELD_NAME;
+  private TrTag renderExistingAdminRow(Http.Request request, long programId, String adminEmail) {
+    return tr().withClasses(
+            ReferenceClasses.ADMIN_TI_GROUP_ROW,
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:130]: TI here means Trusted Intermediary. Since this table is not related to TIs, I don't think we should use this reference class.

File: server/app/views/admin/programs/ManageProgramAdminsView.java [hunk 15]
```diff
@@ -64,72 +71,105 @@ public Content render(
             .setTitle(fullTitle)
             .addMainContent(
                 renderHeader(fullTitle),
-                adminEmailTemplate(),
-                renderAdminForm(request, program.id(), existingAdminEmails));
+                renderHeader("Add New Admin"),
+                renderAddNewButton(request, program.id()),
+                renderHeader("Existing Admins"),
+                renderExistingAdmins(request, program.id(), existingAdminEmails),
+                renderBackButton());
```
Comment (reemaka) [inline server/app/views/admin/programs/ManageProgramAdminsView.java:78]: In most of our other screens, we have the back button at the top of the page

File: server/app/views/admin/programs/ManageProgramAdminsView.java (hunk 10 above)
Comment (yotommy) [reply server/app/views/admin/programs/ManageProgramAdminsView.java:76]: Done

File: server/app/views/admin/programs/ManageProgramAdminsView.java (hunk 15 above)
Comment (yotommy) [reply server/app/views/admin/programs/ManageProgramAdminsView.java:78]: Done

File: server/app/views/admin/programs/ManageProgramAdminsView.java (hunk 11 above)
Comment (yotommy) [reply server/app/views/admin/programs/ManageProgramAdminsView.java:85]: Done

File: server/app/views/admin/programs/ManageProgramAdminsView.java (hunk 12 above)
Comment (yotommy) [reply server/app/views/admin/programs/ManageProgramAdminsView.java:108]: Done

File: server/app/controllers/admin/ProgramAdminManagementController.java (hunk 9 above)
Comment (yotommy) [reply server/app/controllers/admin/ProgramAdminManagementController.java]: Done

File: server/app/views/admin/programs/ManageProgramAdminsView.java (hunk 13 above)
Comment (yotommy) [reply server/app/views/admin/programs/ManageProgramAdminsView.java:104]: Good idea, done.

File: browser-test/image_snapshots/admin_program_admins_test/add-program-admin-error.png
Comment (yotommy) [reply browser-test/image_snapshots/admin_program_admins_test/add-program-admin-error.png:1]: As discussed via Slack, this is difficult because of the need to take on several roles during the test:

1. Log in as admin to create a program
1. Log in as a program admin in order to *create* the program admin as a side effect
1. Log in as admin again in order to further administer program admins. 
 
It looks like calling `logout` in between each session clears the DB. Calling `loginAsProgramAdmin` or `loginAsAdmin` from the logged-in sessions directly doesn't work, because the Dev Tools link isn't available on the particular page where the test currently is. I tried navigating `page.goto(BASE_URL)` to get to the home page where the Dev Tools link is, but that's not working for some reason.

Manual testing shows that this flow works as expected.

**Review by reemaka — APPROVED** [review]
